### ISSUE TITLE
IndexError: list index out of range

### ISSUE THREAD

**Issue by sfesfizh** [requests/requests-ntlm#59]
Hi!

Installed:
`user@note:~> pip freeze`
`python-ntlm3==1.0.2`
`requests==2.7.0`
`requests-ntlm==0.1.0`

My company use web-filter "Websense" with proxy connection and I wrote script for NTLM authentication on "websense":

`#!/usr/bin/python`

`import requests`
`from requests_ntlm import HttpNtlmAuth`

`ip = x.x.x.x`
`host = 'hostname'`
`username = 'domain\user'`
`password = 'password'`
`session = requests.Session()`
`session.auth = HttpNtlmAuth(username,password,session)`
`session.get('http://x.x.x.x:15880/websense?action=login&hostname='+host+'&persistent=1&domain=DOMAIN&ip='+ip+'&ntlm_sess_sec=1')`

But when launch the script I get following error:
`user@note:~> ./websense.py`
`Traceback (most recent call last):`
`  File "./websense.py", line 12, in <module>
    session.get('http://x.x.x.x:15880/websense?action=login&hostname='+host+'&persistent=1&domain=DOMAIN&ip='+ip+'&ntlm_sess_sec=1')`
`  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 477, in get
    return self.request('GET', url, **kwargs)`
`  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 465, in request
    resp = self.send(prep, **send_kwargs)`
`  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 579, in send
    r = dispatch_hook('response', hooks, r, **kwargs)`
`  File "/usr/lib/python2.7/site-packages/requests/hooks.py", line 41, in dispatch_hook
    _hook_data = hook(hook_data, **kwargs)`
`  File "/usr/lib/python2.7/site-packages/requests_ntlm/requests_ntlm.py", line 125, in response_hook
    'Authorization', r, kwargs)`
`  File "/usr/lib/python2.7/site-packages/requests_ntlm/requests_ntlm.py", line 97, in retry_using_http_NTLM_auth`
`    ))[0].strip()`
`IndexError: list index out of range`

Note, when I have installed this configuration:
`python-ntlm3==1.0.2`
`requests==2.2.1`
`requests-ntlm==0.1.0`
it works fine

**Comment by Lukasa** [comment]
So, this bug would happen if for any reason there is no `www-authenticate` header in the second response. There's one other possibility: does it consistently reproduce with requests version 2.7.0 and consistently *not* reproduce with requests version 2.2.1?

**Comment by Lukasa** [comment]
I'm wondering if there is a bug in this block of code:


```python
        ntlm_header_value = list(filter(
            lambda s: s.startswith('NTLM '), auth_header_value.split(',')
        ))[0].strip()
```

**Comment by sfesfizh** [comment]
If any actions I should do?

**Comment by Lukasa** [comment]
@sfesfizh Yeah, can you retest a few times and confirm that you *always* hit this traceback? Or does it work sometimes?

**Comment by sfesfizh** [comment]
@Lukasa , it works always correct in 2.2.1 and no one success in 2.7.0

**Comment by sfesfizh** [comment]
@Lukasa I think error could be with long URL string. Just cropped string to:
`session.get('http://x.x.x.x:15880')`
and script done without any errors

**Comment by Lukasa** [comment]
Long URL doesn't sound right, but I wonder. Are you able to share the URL?

**Comment by sfesfizh** [comment]
@Lukasa yeah, in my first message:
`session.get('http://x.x.x.x:15880/websense?action=login&hostname='+host+'&persistent=1&domain=DOMAIN&ip='+ip+'&ntlm_sess_sec=1')`

**Comment by Lukasa** [comment]
Ok, can you take the failing configuration and slowly increase the requests version until it breaks? I'd like to narrow it down to a specific change log.

**Comment by sfesfizh** [comment]
No problem, please tell me all available versions from 2.2.1.
Thanks!

**Comment by Lukasa** [comment]
2.7.0
2.6.2
2.6.1
2.6.0
2.5.3
2.5.2
2.5.1
2.5.0
2.4.3
2.4.2
2.4.1
2.4.0
2.3.0
2.2.1

**Comment by sfesfizh** [comment]
Hello!

At the next version it comes error:
`requests==2.3.0`

**Comment by Lukasa** [comment]
2.3.0 contained a number of changes, but the relevant ones are probably these:

- No longer expose Authorization or Proxy-Authorization headers on redirect. Fix CVE-2014-1829 and CVE-2014-1830 respectively.
- Authorization is re-evaluated each redirect.

Is it possible for you to use tcpdump or something like it to capture the HTTP transactions and show them to me? All the requests involved (there should be at least two)? Feel free to strip out sensitive information, but please don't remove headers entirely.

**Comment by sfesfizh** [comment]
You need successful responses or not?

**Comment by Lukasa** [comment]
Unsuccessful is preferred. =)

**Comment by sfesfizh** [comment]
Here screens of successfull requests-responses in sequences:
![snapshot1](https://cloud.githubusercontent.com/assets/13018569/8590835/c14a4b1a-262c-11e5-9937-0d92411c6509.png)
![snapshot2](https://cloud.githubusercontent.com/assets/13018569/8590837/c14c51a8-262c-11e5-90b7-ea1b173b731a.png)
![snapshot3](https://cloud.githubusercontent.com/assets/13018569/8590834/c14a0934-262c-11e5-8b2a-010265400d0a.png)
![snapshot6](https://cloud.githubusercontent.com/assets/13018569/8590836/c14c2818-262c-11e5-91db-25bfd0af3ce1.png)
![snapshot4](https://cloud.githubusercontent.com/assets/13018569/8590838/c14dd9ec-262c-11e5-8825-f0469fffdb6c.png)
![snapshot5](https://cloud.githubusercontent.com/assets/13018569/8590839/c14ef71e-262c-11e5-8f3f-437a8188fb42.png)

**Comment by sfesfizh** [comment]
Hm.
While using requests 2.3.0 there are only 2 GET requests and 2 "401" responses, but when using requests 2.2.1 there are 3 GET requests and 2 first "401" responses and 1 "200" response at last.

**Comment by Lukasa** [comment]
@sfesfizh Yeah, I expect there to be one fewer request because the exception prevents us from making the third one.

Are there differences between the second request in both cases?

**Comment by sfesfizh** [comment]
I see differencies in Accept-Encoding: no "compress" and "HTTP request 1/1"
![snapshot7](https://cloud.githubusercontent.com/assets/13018569/8592034/f545886e-2634-11e5-9435-ef2c6335bc39.png)

**Comment by Lukasa** [comment]
Hmm. This is bizarre. Those differences should not be material to the  I wonder if we have a problem with connection setup.

NTLM is really strange and it authenticates connections, not transactions. I wonder: are you getting the same behaviour with TCP connections each time? (A TCP connection is set up by a SYN, SYN-ACK, ACK sequence, and torn down by RST or FIN).

**Comment by sfesfizh** [comment]
Yepp, always 20 packets while in requests == 2.3.0
![snapshot8](https://cloud.githubusercontent.com/assets/13018569/8595109/1c7c855c-2650-11e5-895d-fc468a1b4839.png)

**Comment by Lukasa** [comment]
And to clarify it's exactly the same for 2.2.1?

**Comment by sfesfizh** [comment]
No, on 2.2.1 always 15 packets
![snapshot9](https://cloud.githubusercontent.com/assets/13018569/8595184/b43afcd4-2650-11e5-8020-eddc6043c745.png)

**Comment by Lukasa** [comment]
Aha! Ok, that's likely to be our problem.

In 2.2.1 you've only got 1 TCP connection (from port 47252 to port 15880), but on 2.3.0 you've got two, one per request (from port 47227, and then from port 47228). This breaks NTLM because NTLM auths the individual connection, which you aren't successfully re-using.

I wonder why this is happening. Can you turn on logging to debug level, please, and show me the output in both cases?

**Comment by sfesfizh** [comment]
No problem, but first I want to know what logging do you mean ? :)

**Comment by Lukasa** [comment]
urllib3 does some logging. If you add your code to a script and add the code below to the top of that script you should get some output:

```python
import logging
logging.basicConfig(level=logging.DEBUG)
```

**Comment by sfesfizh** [comment]
Looks like the problem with another connection:

user@note:~> ./websense.py 
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): x.x.x.x
DEBUG:requests.packages.urllib3.connectionpool:"GET /websense?action=login&hostname=note&persistent=1&domain=DOMAIN&ip=x.x.x.x&ntlm_sess_sec=1 HTTP/1.1" 401 0
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (2): x.x.x.x
DEBUG:requests.packages.urllib3.connectionpool:"GET /websense?action=login&hostname=note&persistent=1&domain=DOMAIN&ip=x.x.x.x&ntlm_sess_sec=1 HTTP/1.1" 401 0
Traceback (most recent call last):
  File "./websense.py", line 12, in <module>
    session.get('http://x.x.x.x:15880/websense?action=login&hostname='+host+'&persistent=1&domain=DOMAIN&ip='+ip+'&ntlm_sess_sec=1')
  File "/usr/lib/python2.7/site-packages/requests-2.3.0-py2.7.egg/requests/sessions.py", line 468, in get
    return self.request('GET', url, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests-2.3.0-py2.7.egg/requests/sessions.py", line 456, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python2.7/site-packages/requests-2.3.0-py2.7.egg/requests/sessions.py", line 565, in send
    r = dispatch_hook('response', hooks, r, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests-2.3.0-py2.7.egg/requests/hooks.py", line 41, in dispatch_hook
    _hook_data = hook(hook_data, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests_ntlm/requests_ntlm.py", line 125, in response_hook
    'Authorization', r, kwargs)
  File "/usr/lib/python2.7/site-packages/requests_ntlm/requests_ntlm.py", line 97, in retry_using_http_NTLM_auth
    ))[0].strip()
IndexError: list index out of range

**Comment by Lukasa** [comment]
Hurrah! So the bug is: why is requests using another connection when the one it had should have been good?

**Comment by sfesfizh** [comment]
And now who can correct this bug? :)

**Comment by Lukasa** [comment]
I'm happy to look at this, but I need to find some time first. Definitely won't get there before this evening. =)

**Comment by sfesfizh** [comment]
ok! Whatever, thanks for your help :)

**Comment by Lukasa** [comment]
@sfesfizh Can you try your test with the code from #57? I wonder if this is a symptom of that problem.

**Comment by sfesfizh** [comment]
So I should install requests 2.3.0 and try alternative code for 'requests_ntlm.py'? Right?

**Comment by Lukasa** [comment]
Yeah, please. =)

**Comment by sfesfizh** [comment]
I supose it is all ok :)
Tested with requests 2.3.0 and 2.7.0 and got no error messages, connections dump shows one TCP port communication and three steps of authenticating.

**Comment by Lukasa** [comment]
Alright, I think this means that we need to merge that change.

I'm going to aim to release a new version of requests-ntlm this weekend that contains that fix.

**Comment by sfesfizh** [comment]
Sounds great! )
So I will update new release when it's rolled out :)

Many thanks!

**Comment by Lukasa** [comment]
Ok, #57 is merged, we'll be pushing out 0.2.0 shortly.
